pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -164,13 +165,72 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
 	if t.pollsToArchive <= 0 || t.persister == nil {
 		return
 	}
-	key := fmt.Sprintf("knownFiles%d", t.archiveIndex)
-	if err := checkpoint.SaveKey(context.Background(), t.persister, metadata.Get(), key); err != nil {
+	if err := t.writeArchive(t.archiveIndex, metadata); err != nil {
 		t.set.Logger.Error("error faced while saving to the archive", zap.Error(err))
 	}
 	t.archiveIndex = (t.archiveIndex + 1) % t.pollsToArchive // increment the index
 }
 
+// readArchive loads data from the archive for a given index and returns a fileset.Filset.
+func (t *fileTracker) readArchive(index int) (*fileset.Fileset[*reader.Metadata], error) {
+	key := fmt.Sprintf("knownFiles%d", index)
+	metadata, err := checkpoint.LoadKey(context.Background(), t.persister, key)
+	if err != nil {
+		return nil, err
+	}
+	f := fileset.New[*reader.Metadata](len(metadata))
+	f.Add(metadata...)
+	return f, nil
+}
+
+// writeArchive saves data to the archive for a given index and returns an error, if encountered.
+func (t *fileTracker) writeArchive(index int, rmds *fileset.Fileset[*reader.Metadata]) error {
+	key := fmt.Sprintf("knownFiles%d", index)
+	return checkpoint.SaveKey(context.Background(), t.persister, rmds.Get(), key)
+}
+
+// FindFiles goes through archive, one fileset at a time and tries to match all fingerprints against that loaded set.
+func (t *fileTracker) FindFiles(fps []*fingerprint.Fingerprint) []*reader.Metadata {
+	// To minimize disk access, we first access the index, then review unmatched files and update the metadata, if found.
+	// We exit if all fingerprints are matched.
+
+	mostRecentIndex := (t.archiveIndex - 1 + t.pollsToArchive) % t.pollsToArchive
+	matchedMetadata := make([]*reader.Metadata, len(fps))
+
+	// continue executing the loop until either all records are matched or all archive sets have been processed.
+	for i := 0; i < t.pollsToArchive; i++ {
+		metadataUpdated := false
+
+		// Update the mostRecentIndex
+		currentIndex := mostRecentIndex
+		mostRecentIndex = (mostRecentIndex - 1 + t.pollsToArchive) % t.pollsToArchive
+
+		data, err := t.readArchive(currentIndex) // we load one fileset atmost once per poll
+		if err != nil {
+			t.set.Logger.Error("error while opening archive", zap.Error(err))
+			continue
+		}
+		for index, fp := range fps {
```

Comment:
nit: It's just a loop variable, so `j` seems more idiomatic
```suggestion
		for j, fp := range fps {
```